### ISSUE TITLE
[Nest] Binding - Continuous "Check: Disconnected from streaming events" after network disconnection

### ISSUE THREAD

**Issue by chibbert** [openhab/openhab2-addons#2845]
Hi, firstly apologies if I've filled anything in this bug report out incorrectly, this is my first go at debugging and openhab issue.

I've been using the new nest binding from the 2.2.0-SNAPSHOT.

Randomly, it seems that the binding can get into a state where the no updates are being received by the streaming rest client, and my openhab logs fills up with:

2017-11-14 11:39:01.904 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Receiving streaming events, millisSinceLastEvent=21554
2017-11-14 11:39:10.051 [DEBUG] [nternal.rest.NestStreamingRestClient] - Received 'keep-alive' event, data: null
2017-11-14 11:39:10.052 [DEBUG] [nternal.rest.NestStreamingRestClient] - Received message to keep connection alive
2017-11-14 11:39:31.905 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Receiving streaming events, millisSinceLastEvent=21853
2017-11-14 11:39:40.089 [DEBUG] [nternal.rest.NestStreamingRestClient] - Received 'keep-alive' event, data: null
2017-11-14 11:39:40.090 [DEBUG] [nternal.rest.NestStreamingRestClient] - Received message to keep connection alive
2017-11-14 11:40:01.905 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Receiving streaming events, millisSinceLastEvent=21816
2017-11-14 11:40:31.905 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Receiving streaming events, millisSinceLastEvent=51816
2017-11-14 11:41:01.906 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Disconnected from streaming events, millisSinceLastEvent=81817
2017-11-14 11:41:32.000 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Disconnected from streaming events, millisSinceLastEvent=111911
.......
2017-11-14 11:45:32.142 [DEBUG] [nternal.rest.NestStreamingRestClient] - Check: Disconnected from streaming events, millisSinceLastEvent=352053
and so on.

## Expected Behavior
After a network interruption which causes NestStreamingRestClient to report "Check: Disconnected from streaming events" and go into the state "connected=false" something should restart the connection.

## Current Behavior
After a network interruption which causes NestStreamingRestClient to report "Check: Disconnected from streaming events" and go into the state "connected=false" no reconnection is attempted to restart the connection.

## Possible Solution
Possibly NestStreamingRestAPI could internally break and then re-make the connection.
Or the NestBridgeHandler could occasionally check the connection state of NestStreamIngRestAPI and delete the object and recreate it to handle the reconnection.
I'm not sure which is a nicer solution.

## Steps to Reproduce (for Bugs)
1. Configure the nest binding
2. Check it's working
3. Bounce the network down then up from screen session. e.g:
```
screen
sudo ls
sleep 10; sudo ifdown eno1; sleep 180; sudo ifup eno1
```
now exit screen session to ensure the command keeps running when the network goes down with:
CTL-a
d

Doh, his doesn't reliably reproduce the issue. Turning on and off my broadband router does, but might be tricky to reproduce depending on your setup.

During the time when the connection is down the nest servers will have timed out trying to stream to the binding and the binding will have noticed it is disconnected, but won't do anything about it to reconnect.

## Context
I've been logging through openhab persistence the data from the nest binding to improve the setup of our home heating system.

## Your Environment
* Version used: 2.2.0-SNAPSHOT + manually built nest binding from git master @ Sun Nov 5 10:09:00 
* Environment name and version: openjdk version "1.8.0_152"
* Operating System and version: Ubuntu 16.04.3 LTS, Linux micro 4.4.0-98-generic

**Comment by martinvw** [comment]
@wborn can you assist here?

**Comment by chibbert** [comment]
@wborn I've had a quick go at a fix. I'd appreciate any feedback.
-Fix appears to work for me (rebooting router, and connection is re-established to streamingRestClient)
-I know I need to remove logging, in there for testing at the moment
-I need to check out if I need to change any unit tests too
https://github.com/chibbert/openhab2-addons/tree/2845_nest_binding_reconnect_fix

**Comment by wborn** [comment]
The Jersey EventSource should automatically reconnect on any disconnects (see [documentation](https://jersey.github.io/apidocs/2.22.2/jersey/org/glassfish/jersey/media/sse/EventSource.html)). When it reconnects Nest will again send a `PUT` event with the top level data (see [streaming guide examples](https://developers.nest.com/documentation/cloud/rest-streaming-guide)).

So when there are no programming mistakes in the binding itself, the root cause will probably be an issue in either the Jersey EventSource or the Nest Streaming API implementation.

**Jersey EventSource issues**
I did a quick scan to see if anything related has been reported/fixed in newer versions of Jersey:
* [issue tracker](https://github.com/jersey/jersey/issues?utf8=%E2%9C%93&q=is%3Aissue%20sse)
* [releases](https://github.com/jersey/jersey/releases)
* [release notes](https://jersey.github.io/release-notes) (we are using version 2.22.2)
* [Jersey SSE source code history](https://github.com/jersey/jersey/commits/master/media/sse/src/main/java/org/glassfish/jersey/media/sse)

The only substantial SSE client related fix I could find is in Jersey 2.23 for an issue that fixes closing of a SSE client connection:
* https://github.com/jersey/jersey/issues/3239 ([JERSEY-2967] - Jersey Server-Sent Event Client does not close the connection?)


**Nest Streaming API issues**
The Streaming API is still relatively new so it could still have some issues. There's a report of an [issue](https://nestdevelopers.io/t/no-data-but-keep-alive-over-rest-streaming/887) where the Streaming API keeps sending only keep-alives but no actual event data. I think it is hard to create a workaround for such issues because most of the time keep-alives are being sent while waiting for an actual data event.

**Workaround**
I'm OK with workaround for this issue while we try to get the root cause fixed. From your logging it seems like the Jersey EventSource does not properly reconnect after disconnection otherwise `keep-alive` events would have been received. 

* I think it would be better for the `checkConnectionJob` in `NestStreamingRestClient` to also check and reconnect when it detects that is has been disconnected for while. That way being connected remains the responsibility of the `NestStreamingRestClient`.
* Extra debug logging is good as long as it has some context because it helps debugging issues like these.
* Can you also add a comment in the code on why the workaround is there? People might end up removing it as an optimization because it may seem unnecessary after only reading the EventSource or Nest Streaming API documentation. :-)

**Root cause**
As the root cause of this issue seems to be the Jersey EventSource reconnection logic we could:
1. Test with a newer EventSource version of Jersey to see if it fixes the issue.
2. See if we can increase the Jersey log level and analyze the generated logging when the issue occurs.
3. Create a Jersey issue with reproduction scenario and logging.
4. Fix the issue ourselves (e.g. reproduce issue while step through the code with a debugger, or acquiring a heap dump and explain using the code and object state why the EventSource does not reconnect).
